### PR TITLE
pkg/tinydtls: remove sock_udp_ep_t from sock_dtls_session_t

### PR DESCRIPTION
**Pull request by janosbrodbeck**
### Contribution description
This PR removes the `sock_udp_ep_t` endpoint from the `sock_dtls_session_t` struct in the implementation of tinyDTLS. 
Currently `sock_dtls_session_t` contains a `session_t` and `sock_udp_ep_t` struct which both provide the same information.
```
/* sock_udp_ep_t */
struct _sock_tl_ep {
    int family;
    union {
        uint8_t ipv6[16];
        uint8_t ipv4[4];        /**< IPv4 address mode */
        uint32_t ipv4_u32;      /**< IPv4 address *in network byte order* */
    } addr;                 /**< address */

    uint16_t netif;
    uint16_t port;          /**< transport layer port (in host byte order) */
};

/* session_t */
typedef struct {
    unsigned char size;
    ipv6_addr_t addr;
    unsigned short port;
    int ifindex;
} session_t;
```

`session_t` is the session object from tinyDTLS. By removing the `sock_udp_ep_t` we lower the needed memory size of a `sock_dtls_session_t` object to 24 bytes (upstream: 48 bytes). There are no functional changes or limitations.

A function to retrieve the endpoint from a session will come as a follow up.

### Testing procedure
No errors indicated by Murdock.

If you want to test it on your own: tests/pkg_tinydtls_sock_async.
* Flash the test on two devices (or native).
* Run `dtlss start` on one of them.
* Execute `dtlsc {ip} {data}` from the other node. `{ip}` should contain the IP from the node on which `dtlss start` was executed 

### Issues/PRs references
Resulted from #15615

### CODE REVIEW

File: pkg/tinydtls/contrib/sock_dtls.c [hunk 1]
```diff
@@ -163,13 +163,10 @@ static int _get_psk_info(struct dtls_context_t *ctx, const session_t *session,
     (void)desc;
     (void)desc_len;
     int ret;
-    sock_dtls_session_t _session;
     sock_udp_ep_t ep;
-    sock_dtls_t *sock = (sock_dtls_t *)dtls_get_app_data(ctx);
 
+    sock_dtls_t *sock = (sock_dtls_t *)dtls_get_app_data(ctx);
```
Comment (maribu) [inline pkg/tinydtls/contrib/sock_dtls.c:168]:
```suggestion
    sock_dtls_t *sock = dtls_get_app_data(ctx);
```

*tl;dr*: Try to avoid explicit casts, if possible.

----

`void *` can be implicitly casted to any type in C. (There are some issues with C++ with doing so, so this cannot be done in headers also included by C++. But in C files, implicit casts from `void *` to any other pointer type is fine and even preferred.) The problem with explicit casts is that they turn off all type safety checks.

E.g. consider this:

```C
int main (void) {
    int *a = malloc(sizeof(int) * 32);
    int *b = (int *)malloc(sizeof(int) * 32);
    return 0;
}
```

Because there the `#include <stdlib.h>` is missing, the function `void * malloc(size_t n_bytes)` is undeclared. Ancient C allowed for implicit function declarations but just using them - which still is allowed for backward compatibility unless explicitly turned off. For implicit declarations the argument numbers and types are deduced from the parameters it is called, but the return value is always assumed to be `int`. For `int *a`, the compiler will warn about an `int` being cast to a pointer. For `int *b`, no warning is emitted, as an explicit cast from `int` to `int *` is allowed. The problem is when `sizeof(int) != sizeof(int *)` (e.g. on 64 bit Linux), runtime errors will occur.

This specific issue is not too relevant to RIOT, as we luckily have the compiler flags in place that forbid implicit function declarations. But this is just one of the safety check that get's disabled with explicit casts.

**Review by maribu — COMMENTED** [review]
looks good to me. One minor comment.

**Comment by miri64** [comment]
Did I miss it or are the getters and setters we talked about in #15615 missing?

**Comment by maribu** [comment]
> Did I miss it or are the getters and setters we talked about in #15615 missing?

This seems to be intentionally, if I didn't misinterpret this:

> A function to retrieve the endpoint from a session will come as a follow up.

**Comment by miri64** [comment]
> > Did I miss it or are the getters and setters we talked about in #15615 missing?
> 
> This seems to be intentionally, if I didn't misinterpret this:
> 
> > A function to retrieve the endpoint from a session will come as a follow up.

Okay, I actually missed that sentence :sweat_smile:

File: pkg/tinydtls/contrib/sock_dtls.c [hunk 2]
```diff
@@ -519,6 +516,8 @@ ssize_t sock_dtls_recv_aux(sock_dtls_t *sock, sock_dtls_session_t *remote,
     assert(data);
     assert(remote);
 
+    sock_udp_ep_t ep;
+    _session_to_ep(&remote->dtls_session, &ep);
```
Comment (pokgak) [inline pkg/tinydtls/contrib/sock_dtls.c:520]: `ep` is labelled as `[out]` in the declaration of `sock_udp_recv_aux` so we can skip converting session to ep.

File: pkg/tinydtls/contrib/sock_dtls.c [hunk 3]
```diff
@@ -163,13 +163,10 @@ static int _get_psk_info(struct dtls_context_t *ctx, const session_t *session,
     (void)desc;
     (void)desc_len;
     int ret;
-    sock_dtls_session_t _session;
     sock_udp_ep_t ep;
-    sock_dtls_t *sock = (sock_dtls_t *)dtls_get_app_data(ctx);
 
+    sock_dtls_t *sock = (sock_dtls_t *)dtls_get_app_data(ctx);
     _session_to_ep(session, &ep);
```
[the remark this answers is not in the document]
Comment (pokgak) [reply pkg/tinydtls/contrib/sock_dtls.c:169]: .

[the remark this answers is not in the document]
Comment (janosbrodbeck) [reply pkg/tinydtls/contrib/sock_dtls.c:169]: The same in `get_ecdsa_key`. The session of the parameter is unused. Since it is also a internal function I clean it up.

File: pkg/tinydtls/contrib/sock_dtls.c (hunk 1 above)
Comment (janosbrodbeck) [reply pkg/tinydtls/contrib/sock_dtls.c:168]: Oh didn't know that. Thanks for the detailed explanation!

**Comment by janosbrodbeck** [comment]
Addressed the comments.

> Did I miss it or are the getters and setters we talked about in #15615 missing?

Yeah, like Maribu said I want to do that in a follow up PR since this PR is only an internal optimization of tinyDTLS I wanted to keep it apart.

**Comment by janosbrodbeck** [comment]
I've already squashed since the static-test failed.

**Comment by miri64** [comment]
> I've already squashed since the static-tests failed.

Welp that's more a warning to us maintainers not to merge it by accident. Please only squash when asked by maintainer or if you know its alright with them :-)

**Review by miri64 — APPROVED** [review]
ACK. Code looks good and Murdock confirms that it still compiles. I tested with both `examples/dtls-sock` and `tests/pkg_tinydtls_sock_async` and both still work as expected.
